File: twidere/src/main/java/org/mariotaku/twidere/fragment/BaseListFragment.java
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.fragment;

import android.app.Activity;
import android.app.ListFragment;
import android.content.BroadcastReceiver;
import android.content.ContentResolver;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AbsListView.OnScrollListener;
import android.widget.ListView;

import org.mariotaku.twidere.Constants;
import org.mariotaku.twidere.app.TwidereApplication;
import org.mariotaku.twidere.fragment.iface.RefreshScrollTopInterface;
import org.mariotaku.twidere.util.AsyncTwitterWrapper;
import org.mariotaku.twidere.util.MultiSelectManager;
import org.mariotaku.twidere.util.Utils;

public class BaseListFragment extends ListFragment implements Constants, OnScrollListener, RefreshScrollTopInterface {

	private boolean mActivityFirstCreated;
	private boolean mIsInstanceStateSaved;

	private boolean mReachedBottom, mNotReachedBottomBefore = true;

	public final TwidereApplication getApplication() {
		return TwidereApplication.getInstance(getActivity());
	}

	public final ContentResolver getContentResolver() {
		final Activity activity = getActivity();
		if (activity != null) return activity.getContentResolver();
		return null;
	}

	public final MultiSelectManager getMultiSelectManager() {
		return getApplication() != null ? getApplication().getMultiSelectManager() : null;
	}

	public final SharedPreferences getSharedPreferences(final String name, final int mode) {
		final Activity activity = getActivity();
		if (activity != null) return activity.getSharedPreferences(name, mode);
		return null;
	}

	public final Object getSystemService(final String name) {
		final Activity activity = getActivity();
		if (activity != null) return activity.getSystemService(name);
		return null;
	}

	public final int getTabPosition() {
		final Bundle args = getArguments();
		return args != null ? args.getInt(EXTRA_TAB_POSITION, -1) : -1;
	}

	public AsyncTwitterWrapper getTwitterWrapper() {
		return getApplication() != null ? getApplication().getTwitterWrapper() : null;
	}

	public void invalidateOptionsMenu() {
		final Activity activity = getActivity();
		if (activity == null) return;
		activity.invalidateOptionsMenu();
	}

	public boolean isActivityFirstCreated() {
		return mActivityFirstCreated;
	}

	public boolean isInstanceStateSaved() {
		return mIsInstanceStateSaved;
	}

	public boolean isReachedBottom() {
		return mReachedBottom;
	}

	@Override
	public void onActivityCreated(final Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		mIsInstanceStateSaved = savedInstanceState != null;
		final ListView lv = getListView();
		lv.setOnScrollListener(this);
	}

	@Override
	public void onAttach(final Activity activity) {
		super.onAttach(activity);
	}

	@Override
	public void onCreate(final Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		mActivityFirstCreated = true;
	}

	@Override
	public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
		return super.onCreateView(inflater, container, savedInstanceState);
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		mActivityFirstCreated = true;
	}

	public void onPostStart() {
	}

	@Override
	public void onScroll(final AbsListView view, final int firstVisibleItem, final int visibleItemCount,
			final int totalItemCount) {
		final boolean reached = firstVisibleItem + visibleItemCount >= totalItemCount
				&& totalItemCount >= visibleItemCount;

		if (mReachedBottom != reached) {
			mReachedBottom = reached;
			if (mReachedBottom && mNotReachedBottomBefore) {
				mNotReachedBottomBefore = false;
				return;
			}
			if (mReachedBottom && getListAdapter().getCount() > visibleItemCount) {
				onReachedBottom();
			}
		}

	}

	@Override
	public void onScrollStateChanged(final AbsListView view, final int scrollState) {

	}

	@Override
	public void onStart() {
		super.onStart();
		onPostStart();
	}

	@Override
	public void onStop() {
		mActivityFirstCreated = false;
		super.onStop();
	}

	public void registerReceiver(final BroadcastReceiver receiver, final IntentFilter filter) {
		final Activity activity = getActivity();
		if (activity == null) return;
		activity.registerReceiver(receiver, filter);
	}

	@Override
	public boolean scrollToStart() {
		Utils.scrollListToTop(getListView());
		return true;
	}

	public void setProgressBarIndeterminateVisibility(final boolean visible) {
		final Activity activity = getActivity();
		if (activity == null) return;
		activity.setProgressBarIndeterminateVisibility(visible);
	}

	@Override
	public void setSelection(final int position) {
		Utils.scrollListToPosition(getListView(), position);
	}

	@Override
	public boolean triggerRefresh() {
		return false;
	}

	public void unregisterReceiver(final BroadcastReceiver receiver) {
		final Activity activity = getActivity();
		if (activity == null) return;
		activity.unregisterReceiver(receiver);
	}

	protected void onReachedBottom() {

	}
}

File: twidere/src/main/java/org/mariotaku/twidere/fragment/BaseListFragment.kt
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.fragment

import android.content.ContentResolver
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v4.app.ListFragment
import android.widget.AbsListView
import android.widget.AbsListView.OnScrollListener
import org.mariotaku.twidere.app.TwidereApplication
import org.mariotaku.twidere.constant.IntentConstants.EXTRA_TAB_POSITION
import org.mariotaku.twidere.fragment.iface.RefreshScrollTopInterface
import org.mariotaku.twidere.util.AsyncTwitterWrapper
import org.mariotaku.twidere.util.SharedPreferencesWrapper
import org.mariotaku.twidere.util.Utils
import org.mariotaku.twidere.util.dagger.GeneralComponentHelper
import javax.inject.Inject

open class BaseListFragment : ListFragment(), OnScrollListener, RefreshScrollTopInterface {

    @Inject
    lateinit var twitterWrapper: AsyncTwitterWrapper
    @Inject
    lateinit var preferences: SharedPreferencesWrapper
    var activityFirstCreated: Boolean = false
        private set
    var instanceStateSaved: Boolean = false
        private set
    var reachedBottom: Boolean = false
        private set
    private var notReachedBottomBefore = true

    override fun onAttach(context: Context) {
        super.onAttach(context)
        GeneralComponentHelper.build(context).inject(this)
    }

    val application: TwidereApplication
        get() = TwidereApplication.getInstance(activity)

    val contentResolver: ContentResolver?
        get() {
            val activity = activity
            if (activity != null) return activity.contentResolver
            return null
        }

    fun getSharedPreferences(name: String, mode: Int): SharedPreferences? {
        val activity = activity
        if (activity != null) return activity.getSharedPreferences(name, mode)
        return null
    }

    fun getSystemService(name: String): Any? {
        val activity = activity
        if (activity != null) return activity.getSystemService(name)
        return null
    }

    val tabPosition: Int
        get() {
            val args = arguments
            return if (args != null) args.getInt(EXTRA_TAB_POSITION, -1) else -1
        }

    fun invalidateOptionsMenu() {
        val activity = activity ?: return
        activity.invalidateOptionsMenu()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        instanceStateSaved = savedInstanceState != null
        val lv = listView
        lv.setOnScrollListener(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityFirstCreated = true
    }

    override fun onDestroy() {
        super.onDestroy()
        activityFirstCreated = true
    }

    fun onPostStart() {
    }

    override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int,
                          totalItemCount: Int) {
        val reached = firstVisibleItem + visibleItemCount >= totalItemCount && totalItemCount >= visibleItemCount

        if (reachedBottom != reached) {
            reachedBottom = reached
            if (reachedBottom && notReachedBottomBefore) {
                notReachedBottomBefore = false
                return
            }
            if (reachedBottom && listAdapter.count > visibleItemCount) {
                onReachedBottom()
            }
        }

    }

    override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {

    }

    override fun onStart() {
        super.onStart()
        onPostStart()
    }

    override fun onStop() {
        activityFirstCreated = false
        super.onStop()
    }

    override fun scrollToStart(): Boolean {
        Utils.scrollListToTop(listView)
        return true
    }

    override fun setSelection(position: Int) {
        Utils.scrollListToPosition(listView, position)
    }

    override fun triggerRefresh(): Boolean {
        return false
    }

    protected fun onReachedBottom() {

    }
}
